rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala

```diff
@@ -0,0 +1,679 @@
+      rootNode <- impl.constructNodeFromItem(itemOpt.get)
+
+      notExistingKey = TestData.hexKey("000").toVector
+      readedDataOpt  <- impl.read(rootNode, notExistingKey)
+
+      _ = readedDataOpt.map(readedData => readedData shouldBe none)
+
+    } yield ()
+  }
+
+  "collisions in KVDB" should "be detected" in withImplAndStore { (impl, inMemoStore) =>
+    def copyBVToBuf(bv: ByteVector): ByteBuffer = {
+      val arr    = bv.toArray
+      val newBuf = ByteBuffer.allocateDirect(arr.length)
+      newBuf.put(arr).rewind()
+    }
+
+    val insertRecord    = createInsertActions(List(("FF00FFF01", 0xAA.toByte)))
+    val deleteRecord    = createDeleteActions(List("FF00FFF01"))
+    val collisionKVPair = (copyBVToBuf(History.emptyRootHash.bytes), insertRecord(0).hash.bytes)
+    for {
+      rootNode <- impl.loadNode(RadixHistory.emptyRootHash.bytes, noAssert = true)
+
+      //  process
+      newRootNodeOpt1 <- impl.makeActions(rootNode, insertRecord)
+
+      printedTree1 <- impl.printTree(newRootNodeOpt1.get, "TREE1", true)
+      hashOpt1 <- newRootNodeOpt1.traverse { newRootNode =>
+                   val hash      = impl.saveNode(newRootNode)
+                   val blakeHash = Blake2b256Hash.fromByteVector(hash)
+                   impl.commit.as(blakeHash)
+                 }
+      clrWriteCache = impl.clearWriteCache()
+      printedTree2  <- impl.printTree(newRootNodeOpt1.get, "TREE2", true)
+
+      _ <- inMemoStore.put[ByteVector](Seq(collisionKVPair), copyBVToBuf)
+
+      newRootNodeOpt2 <- impl.makeActions(newRootNodeOpt1.get, deleteRecord)
+      hashOpt         = newRootNodeOpt2.map(node => impl.saveNode(node))
+      err             <- impl.commit.attempt
+
+    } yield {
+      err.isLeft shouldBe true
+      val ex = err.left.get
+      ex shouldBe a[RuntimeException]
+      ex.getMessage shouldBe
+        s"1 collisions in KVDB (first collision with key = " +
+          s"${History.emptyRootHash.bytes.toHex})."
+
+      println(ex.getMessage)
+    }
+  }
+
+  "encode and decode" should "give initial node" in withImplAndStore { (impl, _) =>
+    for {
+      item1 <- impl.update(
+                RadixTree.EmptyItem,
+                TestData.hexKey("FFF8AFF1").toVector,
+                createBV32(0xAD.toByte)
+              )
+
+      node           <- impl.constructNodeFromItem(item1.get)
+      printedTreeStr <- impl.printTree(node, "NODE BEFORE DECODING", true)
+
+      serializedNode = RadixTree.Codecs.encode(node)
+
+      deserializedNode = RadixTree.Codecs.decode(serializedNode)
+
+      printedTreeStr <- impl.printTree(node, "NODE AFTER SERIALIZE", true)
+
+      etalonString = "ByteVector(37 bytes, 0xff03f8aff100000000000000000000000000000000000000000000000000000000000000ad)"
+
+      _ = deserializedNode shouldBe node
+      _ = serializedNode.toString() shouldBe etalonString
+    } yield ()
+
+  }
+
+  "function makeActions" should "not create artefacts" in withImplAndStore { (impl, inMemoStore) =>
```

Comment:
Instead of all these MacActions tests, let's do the 1 test, with this Actions
1 action - With the help of makeactions to build a tree with a set of data as in the example to the specification: 0x111122334455 0x00..0001, 0x11112233aabb 0x00..0002, 0x1111Aabbcc 0x00..0003, 0x33 0x00..0004, 0xFF0011 0x00..0005, 0xFF012222 0x00 ..0006.
2 action - Print this tree and check with etalon
3 action - Validated that the number of nodes in Store is equal to 4
4 action - Delete all data from this tree with help MakeAction()
5 action - print empty tree, compare with etalon
6 action - Validated that the number of nodes in Store is equal to 5 (4 - old data + 1 empty node).